### PR TITLE
Add --no-lap-sync cmd option to ann-bench

### PR DESCRIPTION
**Pull request by achirkin**
Add a command-line option to disable the CUDA event/stream synchronization on each iteration. Only one sync is done per benchmark loop in this case instead. As a result, the measured QPS is observed due to:
1) A small `cudaEventSynchronize` is removed from the benchmark loop;
2) If a GPU algorithm has little to no sync between the GPU and CPU, the kernel launch latency and other CPU overheads are completely hidden.

The new option is experimental and disabled by default.

### CODE REVIEW

**Comment by achirkin** [comment]
NB: this change is required for benchmarking of the dynamic batching PR #261. Without `--no-lap-sync` the search requests cannot huddle up together into batches due to the barriers between the benchmark iterations.

File: cpp/bench/ann/src/common/benchmark.hpp [hunk 1]
```diff
@@ -300,25 +303,28 @@ void bench_search(::benchmark::State& state,
...
-                  n_queries,
-                  k,
-                  neighbors_ptr + out_offset * k,
-                  distances_ptr + out_offset * k);
-      } catch (const std::exception& e) {
-        state.SkipWithError("Benchmark loop: " + std::string(e.what()));
-        break;
-      }
+    {
+      [[maybe_unused]] auto gpu_all = gpu_timer.lap(no_lap_sync);
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.hpp:307]: No, in both cases these are here on purpose. The time is measured between construction and destruction and added to the gpu_time. When `no_lap_sync` is enabled we still need at least one sync around the loop to make sure the GPU has finished its work before we measure the total run time.

File: cpp/bench/ann/src/common/benchmark.hpp [hunk 2]
```diff
@@ -149,9 +150,10 @@ void bench_build(::benchmark::State& state,
   cuda_timer gpu_timer{algo};
   {
     nvtx_case nvtx{state.name()};
+    [[maybe_unused]] auto gpu_all = gpu_timer.lap(no_lap_sync);
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.hpp:153]: Yes, implicitly (see below).

File: cpp/bench/ann/src/common/benchmark.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/bench/ann/src/common/benchmark.hpp:307]: Let's add a note about this: 
> The time is measured between construction and destruction of gpu_all and added to the gpu_time. When no_lap_sync is enabled we still need at least one sync around the loop to make sure the GPU has finished its work before we measure the total run time. This sync is performed by the destructor of gpu_timer.

[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/bench/ann/src/common/benchmark.hpp:319]: Thanks, that is a good point! Added the comment.

File: cpp/bench/ann/src/common/benchmark.hpp [hunk 3]
```diff
@@ -149,9 +150,21 @@ void bench_build(::benchmark::State& state,
   cuda_timer gpu_timer{algo};
   {
     nvtx_case nvtx{state.name()};
+    /* Note: GPU timing
+
+    The GPU time is measured between construction and destruction of `cuda_lap` objects (`gpu_all`
+    and `gpu_lap` variables) and added to the `gpu_timer` object.
+
+    We sync with the GPU (cudaEventSynchronize) either each iteration (lifetime of the `gpu_lap`
+    variable) or once per benchmark loop (lifetime of the `qpu_all` variable). The decision is
```
Comment (tfeher) [inline cpp/bench/ann/src/common/benchmark.hpp:159]:
```suggestion
    variable) or once per benchmark loop (lifetime of the `gpu_all` variable). The decision is

```

**Review by tfeher — APPROVED** [review]
Thanks Artem, LGTM (just on typo in the comments).

**Pull request merged**